syntax: Add labeled block break

Keno opened the pull request:
This is a redo of #60367 taking into account various feedback on that PR. In particular, it seemed like people strongly disliked the `break break` syntax for two primary reasons:

1. It scales poorly to multiple loops `break break break break`
2. It introduced footguns if extra loops are introduced between the loop and the break.

Instead, the consensus seemed to be that labeled break should be the only facility available. Thus, this implements a proper labeled break facility that looks as follows:

```
@label :name for i = 1:10
  for j = 1:10
    break :name (i, j)
  end
end # evaluate to `(1,1)
```

The idea is to re-use the `@label` macro for now, but possibly promote this to syntax in a future version if it gains widespread use. Note that parser changes are still required, since the `break` syntax is currently an error. However, compat.jl could provide `@goto break name val`, which parses fine.

`continue` is extended with label support as well:
```
@label :outer while true
  for i = 1:10
    a[i] && continue :outer
  end
  [...]
end
```

However, the feature is not restricted to loops. A particular use case is to replace cleanup blocks written like

```
cond1 && @goto error
cond2 && @goto error

return true

@label error
error("foo")
```

by a labeled begin/end block:
```
@label :error begin
    cond1 && break :error
    cond2 && break :error
    return true
end
error("foo")
```

This doesn't save much typing work, but it makes this kind of pattern much more structured, e.g. for code-folding in IDEs.

A similar pattern replaces the `for-then` construction originally proposed:

```
result = @label :result begin
  for x in arr
    pred(x) && break :result x
  end
  default
end
```

For convenience, the label may be ommitted, in which case it defaults to `_`, i.e. the above can be written as:

```
result = @label begin
  for x in arr
    pred(x) && break _ x
  end
  default
end
```

I've taken the liberty of converting some base code for testing and to give an idea of what the syntax looks like in practice, but didn't go through particularly comprehensively. These changes should be considered extended usage examples.

Largely written by Claude, and I haven't looked at the implementation particularly carefully yet - for now I'm just interested in discussion of the syntax.

Keno committed:
Add extended block-break syntax

This is a redo of #60367 taking into account various feedback on that
PR. In particular, it seemed like people strongly disliked the `break
break` syntax for two primary reasons:

1. It scales poorly to multiple loops `break break break break`
2. It introduced footguns if extra loops are introduced between the
   loop and the break.

Instead, the consensus seemed to be that labeled break should be
the only facility available. Thus, this implements a proper labeled
break facility that looks as follows:

```
@label name for i = 1:10
  for j = 1:10
    break name (i, j)
  end
end # evaluate to `(1,1)
```

The idea is to re-use the `@label` macro for now, but possibly promote
this to syntax in a future version if it gains widespread use. Note
that parser changes are still required, since the `break` syntax is
currently an error. However, compat.jl could provide `@goto break name
val`, which parses fine.

`continue` is extended with label support as well:
```
@label outer while true
  for i = 1:10
    a[i] && continue outer
  end
  [...]
end
```

However, the feature is not restricted to loops. A particular use
case is to replace cleanup blocks written like

```
cond1 && @goto error
cond2 && @goto error

return true

@label error
error("foo")
```

by a labeled begin/end block:
```
@label error begin
    cond1 && break error
    cond2 && break error
    return true
end
error("foo")
```

This doesn't save much typing work, but it makes this kind of
pattern much more structured, e.g. for code-folding in IDEs.

A similar pattern replaces the `for-then` construction originally
proposed:

```
result = @label result begin
  for x in arr
    pred(x) && break result x
  end
  default
end
```

For anonymous blocks, use `@label _ begin ... end` with `break _`:

```
result = @label _ begin
  for x in arr
    pred(x) && break _ x
  end
  default
end
```

I've taken the liberty of converting some base code for testing and
to give an idea of what the syntax looks like in practice, but didn't
go through particularly comprehensively. These changes should
be considered extended usage examples.

Largely written by Claude.

Keno commented:
Capturing some discussion from this morning:

There appears to be general support for this version of the proposal, so I will move forwards with cleanup and getting this ready to merge. Some specific concerns:

- @StefanKarpinski was concerned that if we want to move to a proper label syntax in the near future, this would introduce a duplicate syntax. However, I don't think any potential first class syntax is necessarily that close - it intersects with thinking about general syntax for scopes/lifetimes, which is not something I want to tackle immediately.

- There was some general skepticism of the omission of the label. I will change this back to require `_` in the `@label` invocation. We can reconsider in the future if it turns out annoying.

- There was some general discussion of whether this syntax obsoletes ordinary `@label`/`@goto`. The answer is that the possible control flow of this feature is more restricted. In particular, the control flow is structured and the resulting CFG will be reducible. This distinction does not currently matter in the current IR representation, but there was some thoughts on making this more explicit in the IR in the future.

bbrehm commented:
> and when targeting the Nth loop from the inside, it’s equivalent to a labeled break of the N-1th loop

Consider:
```
@label_for_continue_break :A while condition1
    #Block1
    while condition2
        #...
         continue :A 
    end
    #Block2
end :A
```
The `continue :A` will skip Block2.

I further like the fact that labeled `begin` with labeled continue supports blocks/loops that don't introduce scope, and are potentially clearer than a `while true` for loops that should run at least once, "do-while"-style.

Keno commented:
Labeled continue is already implemented by this PR

Keno commented:
> @Keno is going to vibecode the proposal with

I implemented this, but I think I will not PR it - I think it's too clever.

I've played with a few more variants, but I keep coming back to the proposal as currently in this PR. I also think we should keep the `:` - I really think that identifier juxtaposition like
```julia
function break_with_underscore()
    @label x begin
        for i in 1:10
            if i > 5
                break x i * 3
            end
        end
        0
    end
end
```
is very jarring on the eyes. That said, maybe that can be mitigated with syntax highlighting - I'll play with that as well.

Keno committed:
Strip 1.14 module markers when comparing parses

This violates a little bit the intent of syntax versioning - which
was to keep flisp at 1.13 and put new features into JuliaSyntax.
However, I do think it's important to use new features in bootstrap
to make sure they get additional testing. Unfortunately, we haven't
yet figured out the bootstrap story for JuliaSyntax, so our flisp
parser is a bit of a hybrid now where it implements some, but not all
1.14 syntax features. I think that's fine for now, in the expectation
that we'll get a bootstrap story for 1.14, but we should re-evaluate.

Keno commented:
> One more question, are there any potential problems with having unused `symbolicblock` expressions in macro output, or will they just go away at some point in the compilation?

Will be gone after lowering.
